src/ngraph/op/fused/fake_quantize.cpp

```diff
@@ -0,0 +1,139 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include <memory>
+
+#include "fake_quantize.hpp"
+#include "ngraph/op/constant.hpp"
+#include "ngraph/op/util/broadcasting.hpp"
+#include "ngraph/shape.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+op::FakeQuantize::FakeQuantize(const shared_ptr<Node>& data,
+                               const shared_ptr<Node>& input_low,
+                               const shared_ptr<Node>& input_high,
+                               const shared_ptr<Node>& output_low,
+                               const shared_ptr<Node>& output_high,
+                               size_t levels)
+    : FusedOp("FakeQuantize", {data})
+    , m_levels(levels)
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::FakeQuantize::pre_validate_and_infer_types()
+{
+    const auto& data_pshape = get_input_partial_shape(0);
+    const auto& input_low_pshape = get_input_partial_shape(1);
+    const auto& input_high_pshape = get_input_partial_shape(2);
+    const auto& output_low_pshape = get_input_partial_shape(3);
+    const auto& output_high_pshape = get_input_partial_shape(4);
+
+    if (data_pshape.is_static() && input_low_pshape.is_static() && input_high_pshape.is_static() &&
+        output_low_pshape.is_static() && output_high_pshape.is_static())
+    {
+        const Shape& data_shape{data_pshape.to_shape()};
```

Comment:
These statements look odd to me. You are taking the reference of the temporary value returned by to_shape(). Where is the temporary stored? I would feed better if this were just a const Shape and not a reference. Does this code work for all of our compilers?

Reply:
I've changed those references to local objects.